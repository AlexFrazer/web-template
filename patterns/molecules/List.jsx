import React, { Component, PropTypes } from 'react';

export default class List extends Component {
  static propTypes = {
    data: PropTypes.arrayOf(PropTypes.shape({
      id: PropTypes.any.isRequired,
    })).isRequired,
    retrieve: PropTypes.func.isRequired,
    children: PropTypes.func.isRequired,
    isFetching: PropTypes.bool.isRequired,
  };

  componentDidMount() {
    this.props.retrieve();
  }

  render() {
    const { data, isFetching, children } = this.props;

    return (
      <div>
        {isFetching && <div />}
        {data.map(children)}
      </div>
    );
  }
}
