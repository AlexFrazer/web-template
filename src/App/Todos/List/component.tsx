import * as React from 'react';
import { clearTodos, deleteTodo } from '../actions';
import AddTodo from './AddTodo';

export interface Props {
  readonly todos: ReadonlyArray<Todo>;
  readonly deleteTodo: typeof deleteTodo;
  readonly clearTodos: typeof clearTodos;
}

export default function List({ todos, clearTodos }: Props) {
  return (
    <div>
      <h1>TO DO</h1>
      <ol>
        {todos.map(({ title }) => (
          <li>{title}</li>
        ))}
      </ol>
      <AddTodo />
      <button onClick={clearTodos}>Clear List</button>
    </div>
  );
}
