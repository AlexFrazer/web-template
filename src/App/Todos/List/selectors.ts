import { createSelector } from 'reselect';
import { AppState } from 'src/reducer';

export const selectTodos = createSelector(
  (state: AppState) => state.todos.todos,
  todos => Object.values(todos),
);
