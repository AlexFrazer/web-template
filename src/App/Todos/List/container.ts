import { connect } from 'react-redux';
import { compose } from 'recompose';
import { AppState } from 'src/reducer';
import { selectTodos } from './selectors';
import { clearTodos, deleteTodo } from '../actions';
import component, { Props } from './component';

const mapStateToProps = (state: AppState) => ({
  todos: selectTodos(state),
});

const mapDispatchToProps = {
  clearTodos,
  deleteTodo,
};

export default compose<Props, {}>(
  connect(
    mapStateToProps,
    mapDispatchToProps,
  ),
)(component);
