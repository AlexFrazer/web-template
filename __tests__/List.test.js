import React from 'react';
import renderer from 'react-test-renderer';
import List from '../patterns/molecules/List';

test('Creates a list that requests data', () => {
  let isFetching = true;
  let data = [{ id: 1 }];
  const retrieve = () => {
    return new Promise(resolve => {

    });
    isFetching = true;
    setTimeout(() => {
      isFetching = false;
      return data;
    }, 200);
  }

  const component = renderer.create(
    <List
      data={data}
      retrieve={retrieve}
      isFetching={isFetching}
    >
      {entry => <div key={entry.id} />}
    </List>
  );

  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
