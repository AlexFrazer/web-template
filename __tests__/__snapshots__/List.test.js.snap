exports[`test Creates a list that requests data 1`] = `
<div>
  <div />
  <div />
</div>
`;
